File: tests/Billing.Tests/FinalizeEngineTests.cs
using Billing.Mediation.Context;
using Billing.Mediation.Model;
using Billing.Mediation.Rating;

namespace Billing.Tests;

/// <summary>The finalize compute core: per-tier settlement over the call's chain (detect → resolve →
/// RateCache → A2Z), cash vs package uom split, multi-tier totals, unrated tiers, and the unanswered /
/// empty-chain edges. Mirrors the MaxRateEngine shape (pure engine, chain handed in).</summary>
public class FinalizeEngineTests
{
    // Plan 7: per-minute 1.0 for prefix 1712; tuple assigns it to SG10 customer for partner 5.
    private static MediationContext Mediation() => new()
    {
        RateCache = RateCache.Build(new Dictionary<int, IReadOnlyDictionary<string, Rate>>
        {
            [7] = new Dictionary<string, Rate> { ["1712"] = new() { Prefix = "1712", RateAmount = 1.0m, IdRatePlan = 7 } },
        }, new DateOnly(2026, 6, 19)),
        RatePlanResolver = RatePlanResolver.Build(new[]
        {
            new RatePlanAssignmentTuple { Id = 1, IdService = 10, AssignDirection = (int)AssignmentDirection.Customer, IdPartner = 5, IdRatePlan = 7 },
        }),
    };

    private static readonly IReadOnlyDictionary<int, Partner> RetailPartner5 =
        new Dictionary<int, Partner> { [5] = new() { IdPartner = 5, PartnerType = 3 } };

    private static FinalizeFacts Facts(string called = "8801712345678", int billsec = 60, bool answered = true) =>
        new("telcobright", "8801999000111", called, ServiceType.Voice, SwitchId: 1, "in", "out", billsec, answered, "uid-1");

    private static FinalizeTierInput Tier(string dbName, TierMode mode, TierReserved? reserved) =>
        new(dbName, PartnerId: 5, Mediation(), RetailPartner5, mode, reserved);

    [Fact]
    public void Cash_tier_settles_to_the_a2z_amount()
    {
        var result = FinalizeEngine.Default().Finalize(Facts(),
            new[] { Tier("telcobright", TierMode.Full, new TierReserved(100, "BDT", 2.0m)) });

        Assert.True(result.Success);
        var s = result.Settlements["telcobright"];
        Assert.Equal(10, s.ServiceGroupId);
        Assert.Equal(1.0m, s.Charged);
        Assert.Equal(1.0m, s.InPartnerCost);
        Assert.Equal(0m, s.PackageAmount);
        Assert.Equal("1712", s.MatchedPrefix);
        Assert.Equal(1.0m, result.TotalCharged);
    }

    [Fact]
    public void Package_tier_settles_to_consumed_minutes()
    {
        var result = FinalizeEngine.Default().Finalize(Facts(),
            new[] { Tier("res_233", TierMode.CustomerOnly, new TierReserved(200, "TF_min", 5.0m)) });

        var s = result.Settlements["res_233"];
        Assert.Equal("TF_min", s.Uom);
        Assert.Equal(1.0m, s.PackageAmount);     // 60s → 1 minute
        Assert.Equal(1.0m, s.Charged);
        Assert.Equal(0m, s.InPartnerCost);
    }

    [Fact]
    public void Multi_tier_settles_each_and_sums_the_total()
    {
        var result = FinalizeEngine.Default().Finalize(Facts(),
            new[]
            {
                Tier("res_233", TierMode.CustomerOnly, new TierReserved(200, "BDT", 2.0m)),
                Tier("telcobright", TierMode.Full, new TierReserved(100, "BDT", 2.0m)),
            });

        Assert.True(result.Success);
        Assert.Equal(2, result.Settlements.Count);
        Assert.Equal(1.0m, result.Settlements["res_233"].Charged);
        Assert.Equal(1.0m, result.Settlements["telcobright"].Charged);
        Assert.Equal(2.0m, result.TotalCharged);
    }

    [Fact]
    public void Unrated_tier_fails_the_call()
    {
        // 8809999999 normalizes to 9999999 — no prefix in plan 7.
        var result = FinalizeEngine.Default().Finalize(Facts(called: "8809999999"),
            new[] { Tier("telcobright", TierMode.Full, new TierReserved(100, "BDT", 2.0m)) });

        Assert.False(result.Success);
        Assert.NotNull(result.Settlements["telcobright"].Error);
        Assert.Equal(0m, result.TotalCharged);
    }

    [Fact]
    public void Empty_chain_is_an_unknown_tenant()
    {
        var result = FinalizeEngine.Default().Finalize(Facts(), Array.Empty<FinalizeTierInput>());
        Assert.False(result.Success);
        Assert.Contains("unknown tenant", result.Error);
        Assert.Empty(result.Settlements);
    }

    [Fact]
    public void Unanswered_call_settles_to_zero_without_error()
    {
        var result = FinalizeEngine.Default().Finalize(Facts(billsec: 0, answered: false),
            new[] { Tier("telcobright", TierMode.Full, new TierReserved(100, "BDT", 0m)) });

        Assert.True(result.Success);
        Assert.Equal(0m, result.Settlements["telcobright"].Charged);
        Assert.Equal(0m, result.TotalCharged);
    }
}

File: java/src/test/java/com/telcobright/billing/mediation/rating/FinalizeEngineTests.java
package com.telcobright.billing.mediation.rating;

import com.telcobright.billing.mediation.context.MediationContext;
import com.telcobright.billing.mediation.model.AssignmentDirection;
import com.telcobright.billing.mediation.model.Partner;
import com.telcobright.billing.testsupport.TestData;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * The finalize compute core over the legacy entities: per-tier settlement (detect -&gt; resolve -&gt;
 * PrefixMatcher -&gt; A2ZRater), cash vs package uom split, multi-tier totals, unrated tiers, and the
 * unanswered / empty-chain edges.
 */
class FinalizeEngineTests {
    private static MediationContext Mediation() {
        var f = TestData.fixture();
        f.tup(10, AssignmentDirection.Customer.value, 5, null, 0, TestData.Ra(8801712, "1.0").idRatePlan(7));
        return f.mediation();
    }

    // adds a SG10 supplier tuple (out-partner 7 @ 2.0/min) on top of the customer tuple.
    private static MediationContext MediationWithSupplier() {
        var f = TestData.fixture();
        f.tup(10, AssignmentDirection.Customer.value, 5, null, 0, TestData.Ra(8801712, "1.0").idRatePlan(7));
        f.tup(10, AssignmentDirection.Supplier.value, 7, null, 0, TestData.Ra(8801712, "2.0").idRatePlan(8));
        return f.mediation();
    }

    private static final Map<Integer, Partner> RetailPartner5 = Map.of(5, new Partner(5, null, 3));

    private static FinalizeFacts Facts(String called, int billsec, boolean answered, int outPartnerId) {
        return new FinalizeFacts("telcobright", "8801999000111", called, ServiceType.Voice, 1, "in", "out",
                outPartnerId, LocalDateTime.of(2026, 6, 19, 0, 0, 0), billsec, answered, "uid-1");
    }

    private static FinalizeFacts Facts() {
        return Facts("8801712345678", 60, true, 0);
    }

    private static FinalizeFacts Facts(String called) {
        return Facts(called, 60, true, 0);
    }

    private static FinalizeFacts Facts(int outPartnerId) {
        return Facts("8801712345678", 60, true, outPartnerId);
    }

    private static FinalizeFacts Facts(int billsec, boolean answered) {
        return Facts("8801712345678", billsec, answered, 0);
    }

    private static FinalizeTierInput Tier(String dbName, TierMode mode, TierReserved reserved) {
        return new FinalizeTierInput(dbName, 5, Mediation(), RetailPartner5, mode, reserved);
    }

    @Test
    void Cash_tier_settles_to_the_a2z_amount() {
        var result = FinalizeEngine.Default().Finalize(Facts(),
                List.of(Tier("telcobright", TierMode.Full, new TierReserved(100, "BDT", new BigDecimal("2.0")))));

        assertTrue(result.Success());
        var s = result.Settlements().get("telcobright");
        assertEquals(10, s.ServiceGroupId());
        assertEquals(10, s.ServiceFamilyId());
        assertEquals(0, new BigDecimal("1.0").compareTo(s.Charged()));
        assertEquals(0, new BigDecimal("1.0").compareTo(s.InPartnerCost()));
        assertEquals(0, BigDecimal.ZERO.compareTo(s.PackageAmount()));
        assertEquals("8801712", s.MatchedPrefix());
        assertEquals(0, new BigDecimal("1.0").compareTo(result.TotalCharged()));
    }

    @Test
    void Package_tier_settles_to_consumed_minutes() {
        var result = FinalizeEngine.Default().Finalize(Facts(),
                List.of(Tier("res_233", TierMode.CustomerOnly, new TierReserved(200, "TF_min", new BigDecimal("5.0")))));

        var s = result.Settlements().get("res_233");
        assertEquals("TF_min", s.Uom());
        assertEquals(0, new BigDecimal("1.0").compareTo(s.PackageAmount()));     // 60s -> 1 minute
        assertEquals(0, new BigDecimal("1.0").compareTo(s.Charged()));
        assertEquals(0, BigDecimal.ZERO.compareTo(s.InPartnerCost()));
    }

    @Test
    void Multi_tier_settles_each_and_sums_the_total() {
        var result = FinalizeEngine.Default().Finalize(Facts(),
                List.of(
                        Tier("res_233", TierMode.CustomerOnly, new TierReserved(200, "BDT", new BigDecimal("2.0"))),
                        Tier("telcobright", TierMode.Full, new TierReserved(100, "BDT", new BigDecimal("2.0")))));

        assertTrue(result.Success());
        assertEquals(2, result.Settlements().size());
        assertEquals(0, new BigDecimal("2.0").compareTo(result.TotalCharged()));
    }

    @Test
    void Unrated_tier_fails_the_call() {
        // 8809999999 normalizes to 9999999 — no matching rateassign prefix.
        var result = FinalizeEngine.Default().Finalize(Facts("8809999999"),
                List.of(Tier("telcobright", TierMode.Full, new TierReserved(100, "BDT", new BigDecimal("2.0")))));

        assertFalse(result.Success());
        assertNotNull(result.Settlements().get("telcobright").Error());
        assertEquals(0, BigDecimal.ZERO.compareTo(result.TotalCharged()));
    }

    @Test
    void Empty_chain_is_an_unknown_tenant() {
        var result = FinalizeEngine.Default().Finalize(Facts(), List.of());
        assertFalse(result.Success());
        assertTrue(result.Error().contains("unknown tenant"));
        assertTrue(result.Settlements().isEmpty());
    }

    @Test
    void Full_tier_also_charges_the_supplier_leg() {
        var tier = new FinalizeTierInput("telcobright", 5, MediationWithSupplier(), RetailPartner5,
                TierMode.Full, new TierReserved(100, "BDT", new BigDecimal("5.0")));

        var result = FinalizeEngine.Default().Finalize(Facts(7), List.of(tier));

        var s = result.Settlements().get("telcobright");
        assertEquals(0, new BigDecimal("1.0").compareTo(s.Charged()));        // customer leg (60s @ 1.0/min)
        assertEquals(0, new BigDecimal("2.0").compareTo(s.SupplierCost()));   // supplier leg (60s @ 2.0/min, out-partner 7)
    }

    @Test
    void Customer_only_tier_has_no_supplier_cost() {
        var result = FinalizeEngine.Default().Finalize(Facts(7),
                List.of(Tier("res_233", TierMode.CustomerOnly, new TierReserved(100, "BDT", new BigDecimal("5.0")))));
        assertEquals(0, BigDecimal.ZERO.compareTo(result.Settlements().get("res_233").SupplierCost()));
    }

    @Test
    void Unanswered_call_settles_to_zero_without_error() {
        var result = FinalizeEngine.Default().Finalize(Facts(0, false),
                List.of(Tier("telcobright", TierMode.Full, new TierReserved(100, "BDT", BigDecimal.ZERO))));

        assertTrue(result.Success());
        assertEquals(0, BigDecimal.ZERO.compareTo(result.Settlements().get("telcobright").Charged()));
        assertEquals(0, BigDecimal.ZERO.compareTo(result.TotalCharged()));
    }
}
